Kp 7440 metax create put data

kaisakui opened the pull request:
This PR includes following implementations:
- Records are fetched from OAI-PMH since last harvest date from a log
- PUT and POST requests to Metax based on above
- The CLI has been removed as I didn't find it very useful
- Tests are updated based on all the changes. The tests in test_metadata_harvesting_cli.py do not work and some have been commented out until at least one is implemented successfully.

kaisakui committed:
add metax api

kaisakui committed:
KP-7440 add method to check if dataset is in data-catalog

kaisakui committed:
KP-7440 Get id of dataset from Metax

kaisakui committed:
KP-7440 Make post request to Metax for new dataset

kaisakui committed:
KP-7440 Make put request to Metax for existing dataset

kaisakui committed:
KP-7440 Enable logging requests

kaisakui committed:
KP-7440 Log failed and successful dataset post requests

kaisakui committed:
KP-7440 Log failed and successful dataset put requests

kaisakui committed:
KP-7440 Limit api calls to ignore deleted records

kaisakui committed:
KP-7440 Remove checking if metadata exists

kaisakui committed:
KP-7440 Remove click

I don't see us using this harvester as a cli tool.

kaisakui committed:
KP-7440 Define kielipankki url as default

kaisakui committed:
KP-7440 Remove underscore since method is needed outside the class

kaisakui committed:
KP-7440 Changes methods to be used outside the class

kaisakui committed:
KP-7440 Enable returning a dictionary that maps converted data to a pid

kaisakui committed:
KP-7440 Fix error logging

kaisakui committed:
KP-7440 Fix typos in licenses

kaisakui committed:
KP-7440 Don't return json but dictionary

kaisakui committed:
KP-7440 Change method name and docstring

kaisakui committed:
KP-7440 Change variable names to be more accurate

kaisakui committed:
KP-7440 Fix docstring

kaisakui committed:
KP-7440 Fix changed method name

kaisakui committed:
KP-7440 Add function to check updated records

kaisakui committed:
KP-7440 Add function to put/post data

kaisakui committed:
KP-7440 Change data catalog in dummy data

kaisakui committed:
KP-7440 Add basics for metax api testing

kaisakui committed:
KP-7440 Add tests for metax api requests

kaisakui committed:
KP-7440 Change return statement to return false also

kaisakui committed:
KP-7440 Rename test file

kaisakui committed:
KP-7440 Rename tests according to changed methods

kaisakui committed:
KP-7440 Use new method for checking missing pid

kaisakui committed:
KP-7440 Remove renamed file

kaisakui committed:
KP-7440 Remove deleted records test

The OAI PMH request was changed to ignore deleted records

kaisakui committed:
KP-7440 Change the way corpus type is checked

kaisakui committed:
KP-7440 Fix changed data catalog id

kaisakui committed:
KP-7440 Use a specific logger for metax requests

kaisakui committed:
KP-7440 Update harvester description

kaisakui committed:
KP-7440 Add a harvester specific logger

kaisakui committed:
KP-7440 Enable getting the last successful harvest date

kaisakui committed:
KP-7440 Add functionality to fetch changed records

kaisakui committed:
KP-7440 Try harvesting only changed records

kaisakui committed:
KP-7440 Try sending data to Metax

kaisakui committed:
KP-7440 Try running the pipeline, log when successfull

kaisakui committed:
KP-7440 remove artificial date retrieving

kaisakui committed:
KP-7440 add params and return to docstring

kaisakui committed:
KP-7440 update docstring

kaisakui committed:
KP-7440 Fetch data according existing or non-existing date

kaisakui committed:
KP-7440 Remove redundant function

kaisakui committed:
KP-7440 Fix doctsring param

kaisakui committed:
KP-7440 Ignore deleted items when fetching data

kaisakui committed:
KP-7440 Update doctring

kaisakui committed:
KP-7440 Update doctrings with params and returns

kaisakui committed:
KP-7440 Add tests for getting harvest date from log

kaisakui committed:
KP-7440 Ignore log files

kaisakui committed:
KP-7440 REmove redundant argument

kaisakui committed:
KP-7440 Fix log date checking

kaisakui committed:
KP-7440 Remove click related code

kaisakui committed:
KP-7440 Attempt taking sample xml to sample response

Also contains uncommented code to focus n one test case.

aajarven commented on harvester/metadata_parser.py:
```diff
@@ -62,14 +62,13 @@ def _get_date(self, xpath):
             raise ValueError("No date found")
 
-    def _check_metadatainfo_exists(self):
+    def check_pid_exists(self):
```
This is a nitpick without direct effect on the software working or not, but I would rename this function to "has_pid" because that will result in more natural language like code: `if record.has_pid()` vs `if record.check_pid_exists()`. 

Same for `if resource.check_resourcetype_corpus()` -> `if resource.is_corpus()`.

Edit: now that I'm on the topic, these could even be properties, eliminating the empty parenthesis and highlighting the fact that all the functions do is indeed return a value.

kaisakui replied:
Sounds good. Let's do it later.

kaisakui replied:
Created https://jira.eduuni.fi/browse/KP-7538 where this could be done.

aajarven commented on harvester/metadata_parser.py:
```diff
@@ -41,7 +41,7 @@ def _get_text_xpath(self, xpath):
         return self.xml.xpath(xpath, namespaces={"info": "http://www.ilsp.gr/META-XMLSchema"})[0]
 
-    def _get_identifier(self, xpath):
+    def get_identifier(self, xpath):
```
The change I'm requesting here is not from the lines edited in this PR, so this can be ignored and fixed Later™ if you don't feel like changing things now.

This and other functions using `_get_text_xpath` are dependent on the user knowing that their xpath must contain the `text()` part (otherwise the function returns an `Element` instead of text, which will break stuff if we try to e.g. `urlparse` the element class).

To make the function do what it promises, I would do something like
```
def text_from(xpath):
    elements = self.xml.xpath(xpath, namespaces={...})
    if not elements:
        raise ElementNotFoundError("Element with xpath '%s' not found", xpath)
    return elements[0].text
```
which would be used like `record.text_from("//info:identificationInfo/info:identifier")`, eliminating the need to pepper all our xpaths with the text function.

The custom exception is a "nice to have" at most, but as we currently have
```
try:
    ....
except:
    raise
```
in the CLI, I suppose that we intend to handle some errors somehow.

aajarven replied:
And now that I'm on a topic of simplifying the pre-existing code, I also noticed that we have identical `get_identifier("//info:identificationInfo/info:identifier/text()")` calls but no other use for the `get_identifier` method. If that's the only identifier we are or will be interested in, we could have
```
@property
def pid(self):
    pid_url = urlparse(self.text_from("//info:identificationInfo/info:identifier"))
    return pid_url.netloc + pid_url.path
```
which would limit the information about the location of the identifier to one place.

aajarven replied:
All in all I would perhaps consider if the `get` prefixes could be dropped or replaced. Now they look like Java getters without actually being like them (as they are somewhat more complex than just `public string get_name() { return name; }`).

I have found it a nice rule of thumb that functions that return a thing without altering the state of the system have noun names (e.g. `record.license_url`) and functions that alter the state of the system are verbs (like `add_something` or `sort` for in-place sorting). NB: like all good rules, this does have a lot of exceptions. E.g. when doing expensive operations from the web or disk or whatnot, it often makes sense to do `fetch_some_data` instead of `some_data` to let the user know that there's abnormal cost.

We also seem to treat the class not so much as a parser, but as a representation of the xml document. At least based on the fact that we always do `metadata_record = MSRecordParser(lxml_record)`, `record = MSRecordParser(single_record)` or such as opposed to `metadata_parser = MSRecordParser(whatnot)`. This suggests that we could rename the class to `MSRecord` (maybe even `MetashareRecord` now that we save the characters from the "parser" part). Combining this with the noun naming for the "getters" would give us the nicely readable `record = MetashareRecord(whatnot)` + `record.attribute`.

kaisakui replied:
Definitely good points. I suggest these improvements/changes are added to the backlog instead of handling them in this ticket.

kaisakui replied:
https://jira.eduuni.fi/browse/KP-7538 created for this.

aajarven commented on harvester/metadata_parser.py:
```diff
@@ -186,34 +186,29 @@ def _map_access_rights(self):
-                    "issued": self._get_date("//info:metadataInfo/info:metadataCreationDate/text()"),
-                    "access_rights": self._map_access_rights(),
-                    
+    def data_converter(self):
```
As this returns the dictionary representation of the (interesting parts of) the xml record, it would be intuitive to have the name be something like `to_dict`, as `data_converter` sounds like it would return a data converter (i.e. something that can be used to convert the data, as opposed to the data that has already been converted).

kaisakui replied:
Yup, totally agree. At first I didn't realize that we don't even need to convert to JSON in order to push the data to Metax. It's enough that it's a dictionary. Then I just quickly renamed the method without too much thinking. Let's do this change also in another ticket?

kaisakui replied:
Changed it now.

aajarven commented on harvester/metadata_parser.py:
```diff
@@ -186,34 +186,29 @@ def _map_access_rights(self):
 
-                return json.dumps(output)
+        }
+        return output
```
If we don't expect to be doing post-processing for the dict any more, we could just go with `return {...}`

kaisakui replied:
Will change.

aajarven commented on metadata_harvester_cli.py:
```diff
@@ -1,30 +1,80 @@
+        logger_harvester.info("Success")
+    except:
+        raise
+        
```
I would expect a "CLI" module to be pretty much limited to what's in this `if`: a command line interface to functionality that is found elsewhere. Plus `--help` and argument parsing and other CLI specific stuff if needed, of course. Now we have all kinds of "actual funcitonality" here too. If we will never use the software from other Python programs, only from the command line, then it's not critical, but if we anticipate importing this functionality elsewhere, it would be especially good to relocate the actual functionality functions to other modules to avoid having to import a "cli".

kaisakui replied:
Not entirely sure I got what you meant here so probably better to discuss in person. And sounds like another improvement to be ticketed?

kaisakui replied:
https://jira.eduuni.fi/browse/KP-7538

aajarven commented on metadata_harvester_cli.py:
```diff
@@ -1,30 +1,80 @@
+    :return: date in last line
+    """
+    log_file_path = "harvester.log"
+    with open(log_file_path, "r") as file:
```
This will raise a `FileNotFoundError` or something like that if the file isn't present at all, right? Currently we don't seem to be prepared for that when using this function, so we might want to catch the error already in here and just return `None` (or whatever makes sense).

kaisakui replied:
If retrieve_metadata_content() doesn't find the file, it will move on to fetching all available records from the api. We can raise the error if you think it'd be better that way.

kaisakui replied:
Implemented the FileNotFoundError because during testing, the code couldn't progress without handling the error correctly.

kaisakui committed:
Simplify return statement

kaisakui committed:
KP-7440 Rename method to to_dict()

kaisakui committed:
KP-7440 Handle missing file

kaisakui committed:
KP-7440 Remove resumptionToken

kaisakui committed:
KP-7440 Fix failing tests for retrieve_metadata_content()

aajarven commented on metadata_harvester_cli.py:
```diff
@@ -1,30 +1,83 @@
\ No newline at end of file
+    try:
+        send_data_to_metax(retrieve_metadata_content())
+        logger_harvester.info("Success")
```
The timestamp from this log is what we use to determine the time range of the next update, right? In that case, the steps are something like
1. fetch updated bindings from Metashare
2. update their data in Metax
3. log the current time

which means that if the metadata in Metashare updates during step 2, those changes have occurred before the timestamp in the log file and they will never be written to Metax.

To mitigate, we have two options:
1. log the start time
2. fetch updated bindings from Metashare (since the start time before the last success log)
2. update their data in Metax
4. log success

or

1. store start time (e.g. as a variable)
2. fetch updated bindings from Metashare
2. update their data in Metax
4. log the success and the start time from step 1

kaisakui replied:
Implemented this now so that it logs the start time of a harvest. The code goes through the log until it finds a successful harvest and then uses that harvest's start time for the next round of harvesting.

aajarven replied:
Nice, looks good!

Now we parse the log separately for logging purpose and for the actual record fetching. This is a bit risky, as it's very easy to change one of the hard-coded file names without remembering to change the other one, resulting in logs saying one thing ("updated since forever") and the actual harvesting having done something else ("updated since last week") which can be hard to debug. Now that we already have the datetime at hand, I suggest we pass it directly to `records_to_dict`.

aajarven commented on tests/test_metax_api.py:
```diff
@@ -0,0 +1,124 @@
+
+@pytest.fixture
+def headers():
+    return {'Content-Type': 'application/json'}
```
This fixture seems to be unused

kaisakui replied:
Removed.

aajarven commented on tests/test_metax_api.py:
```diff
@@ -0,0 +1,124 @@
+import metax_api
+import logging
+
+@pytest.fixture(autouse=True)
+def prevent_online_http_requests(monkeypatch):
+    """
+    Patch urlopen so that all non-patched requests raise an error.
+    """
+
+    def urlopen_error(self, method, url, *args, **kwargs):
+        raise RuntimeError(
+            f"Requests are not allowed in tests, but a test attempted a "
+            f"{method} request to {self.scheme}://{self.host}{url}"
+        )
+
+    monkeypatch.setattr(
+        "urllib3.connectionpool.HTTPConnectionPool.urlopen", urlopen_error
+    )
```
Now the same fixture is found in two test modules. If we put it in `conftest.py` at the root of the tests directory, it can (and with `autouse=True` will) be used by all other test modules.

kaisakui replied:
Created conftest.py for shared fixtures.

aajarven commented on tests/test_metadata_parser.py:
```diff
@@ -215,8 +203,8 @@ def test_accesstype_open_record(accesstype_open_record):
-def test_missing_pid_record(missing_pid_record):
+def test_check_pid_exists(missing_pid_record):
     missing_pid_record = MSRecordParser(missing_pid_record)
-    result = missing_pid_record.json_converter()
-    expected_result = None
+    result = missing_pid_record.check_pid_exists()
+    expected_result = False
     assert result == expected_result
```
It would be more _Pythonic_ :monocle_face: to do `assert not missing_pid_record.json_converter()` or `assert missing_pid_record.json_converter() is False`. We need latter if we want the result to be specifically `False` and not any falsy value such as `0` or `[]` or `None`, otherwise the former is better.

The same goes for other `assert whatnot == True` and `assert whatnot == False` lines too. If you want these automatically identified as you type, at least [pylint](https://pypi.org/project/pylint/) can do that and is very very likely available as a vscode plugin (or whatever they are called).

kaisakui replied:
Pylint suggested `assert not missing_pid_record.check_pid_exists()` so going with that for now.

aajarven commented on tests/test_metadata_parser.py:
```diff
@@ -54,7 +54,7 @@ def test_get_description_not_empty(single_record):
 
 def test_get_identifier_not_empty(single_record):
```
This and the other `test_whatnot_not_empty` tests don't tell that much about whether the code in question works, as this test would pass if the identifier was returned as `None`, `"foo"` or real but wrong metadata (say, the title of the data in question). If it's not very hard, I would make these verify the actual result. Or at least part of it (e.g. `123456789 in identifier` or checking some values from a dict if the dict is large).

kaisakui replied:
It's a silly test for all of them, you're right. Some time after creating these tests, we discussed that the XML would not even contain empty elements. Removed these tests.

aajarven commented on tests/test_metax_api.py:
```diff
@@ -0,0 +1,124 @@
+        dataset_id = metax_api.update_dataset(metax_dataset_id, sample_dict)
+
+    assert dataset_id == "000-000-000"
+    assert "Updated dataset" in caplog.text
```
The more idiot-proof way of testing that a data set has indeed been added to Metax (now it could be that we do all the computations but never call the API, or our request is somehow malformed or missing data, or ...) is to inspect the [request history](https://requests-mock.readthedocs.io/en/latest/history.html): minimally checking that the mocker corresponding to the expected endpoint has been called, if necessary/useful also checking that this has happened the expected times and/or with the expected payload.

kaisakui replied:
I tried to apply these suggested tests for all API calls. Could be they are still not right but really hoping they are!

aajarven replied:
In most if not all of these cases we can simplify the assertions by just checking that the mocker has been called, this type of pattern:
```
(.venv) ajarven@mustavaris:~/tmp/mockerexample$ cat program.py 
import requests


def fetch_cat_fact():
    response = requests.get("https://catfact.ninja/fact")
    return response.json()["fact"]


if __name__ == "__main__":
    print(fetch_cat_fact())
```
```
(.venv) ajarven@mustavaris:~/tmp/mockerexample$ cat test_program.py 
import pytest
import requests_mock

from program import fetch_cat_fact


@pytest.fixture
def mock_cat_api():
    with requests_mock.Mocker() as mocker:
        mocker.get(
            "https://catfact.ninja/fact",
            json={"fact": "cats have whiskers", "length": 123},
        )
        yield mocker


def test_fetch_cat_fact(mock_cat_api):
    assert fetch_cat_fact() == "cats have whiskers"
    assert mock_cat_api.called
    assert mock_cat_api.called_once
    assert mock_cat_api.call_count == 1
```
(except we of course don't need all the three asserts, I just wanted to include the most common flavours in the example)

Because we already mock a specific URL, the test will fail if the code queries the wrong endpoint. E.g. if I change the cat fact URL from the example, I get
```
E       requests_mock.exceptions.NoMockAddress: No mock address: GET https://catfact.ninja/fact/or/something
```

When testing queries with payload that might or might not be sent correctly, we do have use for `request_history` though: with it we can access the payload of the request and check that there are as many metadata records as we expect and/or their content is right.

aajarven commented on metadata_harvester_cli.py:
```diff
@@ -1,30 +1,83 @@
+                    pid = metadata_record.get_identifier("//info:identificationInfo/info:identifier/text()")
+                    all_mapped_data_dict[pid] = metadata_record.to_dict()
+        return all_mapped_data_dict
+    except:
+        raise
```
```
try:
    do_something()
except:
    raise
```
works exactly the same (at least if there's no edge cases I'm not familiar with) as 
```
do_something
```
so I think we can remove this kind of try-excepts without any error handling.

kaisakui replied:
Removed.

aajarven reviewed (commented):
Looks pretty nice, not too far from done! There's the timestamp issue I mentioned in daily and I also made some comments, often about removing unnecessary stuff but also some test things. But I suppose that none of those will be very much work.

For things that are not currently tested but maybe should be, my list is:
* The `doc_unstruct_element` version of license is not present in our test data. As this looks like code that is easy to get wrong (especially if that type of license info is not too common), it would be good to have a test case.
* We don't test `get_all_metadata_records` with `limit`, but on the other hand we don't seem to use it either. If it's dead code, I think we should remove it, or if it's going to be used soon, then we should test it.
* We don't test the CLI as a whole, nor the `send_data_to_metax` part of it. If the CLI was just for easy testing or such, this wouldn't be too bad, but given that it seems that that's the intended way to use the tool (as there's no other way to do the whole chain of updating the metadata and logging the completion) so I'd say that we should test that it behaves as expected.

As a hint for future, test coverage can give some idea about things that our tests don't currently cover. Doing
```
$ pip install pytest-cov  # for regular use, this would of course be in requirements_dev.txt
$ python -m pytest --cov --cov-report=term-missing
```
will provide an output that lets you know which lines were not executed at all when running tests. The numbers that come out aren't that useful without some inspection though, as line coverage is a pretty crude metric: it doesn't always (maybe even usually) make sense to strive for 100% line coverage, and the fact that the line has been executed doesn't mean that it has been tested for anything but "this doesn't crash the whole program (at least not with the parameters used in the test)". But nonetheless, it can provide some guidance.

kaisakui committed:
KP-7440 Fetach records since a date and time

kaisakui committed:
KP-7440 Fetch since last successful date time

kaisakui committed:
KP-7440 Add log as an argument when fetching data

kaisakui committed:
KP-7440 Add test for several unsuccesful harvests

kaisakui committed:
KP-7440 Create conftest to share common fixtures

kaisakui committed:
KP-7440 Add doctrings to fixtures and functions

kaisakui committed:
KP-7440 Remove useless tests

kaisakui committed:
KP-7440 remove try-excepts

kaisakui committed:
KP-7440 License in doc_unstruct element

kaisakui committed:
KP-7440 Test get license from documentUnstructured

kaisakui committed:
KP-7440 Fix api tests with updated docstrings

kaisakui committed:
KP-7440 Remove unused fixture

kaisakui committed:
KP-7440 Rename API methods with docstring updates

kaisakui committed:
KP-7440 Rename function

kaisakui committed:
KP-7440 Add missing docstrings

kaisakui committed:
KP-7740 Rename function

kaisakui committed:
KP-7440 Pytest-cov

kaisakui committed:
KP-7440 Rename sample

kaisakui committed:
KP-7440 Update gitignore

kaisakui committed:
KP-7440 Add proper json for put and post requests

kaisakui committed:
KP-7440 Apply pylint suggested changes

kaisakui committed:
KP-7440 Update new catalog id

kaisakui committed:
KP-7440 Update changed search parameters

kaisakui committed:
KP-7440 Apply pylint suggested changes

kaisakui committed:
KP-7440 Fix headers

kaisakui committed:
KP-7440 Move Metax request fixtures

kaisakui committed:
KP-7440 Apply changes of new search parameter

kaisakui committed:
KP-7440 Raise error

kaisakui committed:
KP-7440 Apply pylint suggested changes

kaisakui committed:
KP-7440 Change metax search parameter

kaisakui committed:
KP-7440 Apply some pylint suggested changes

kaisakui committed:
KP-7440 Remove incorrect test

kaisakui committed:
KP-7440 Unify quotes

kaisakui committed:
KP-7440 Apply pylint formatting

kaisakui commented:
I think I addressed all improvement areas. Creating tests for the whole pipeline is not done - tried it but failed miserably. Otherwise, I used pylint to identify improvement areas, same for pytest-cov. Didn't yet apply all found issues but part of them.

aajarven committed:
KP-7440 Ignore vim tempfiles

aajarven committed:
KP-7440 Extract request mocker creation to own fixture

If we need to mock multiple URLs/methods, and wish to do that using
separate fixtures, they all need to share the same Mocker object.
The new `shared_request_mocker` fixture provides that.

The other fixtures return the shared Mocker. This means that in tests
that use multiple mocking fixtures, all of them return the same Mocker
object (i.e. the request history etc is shared across them).

Co-authored-by: Kaisa Kuivalahti <kaisakui@users.noreply.github.com>
Co-authored-by: Sam Hardwick <Traubert@users.noreply.github.com>

aajarven committed:
KP-7440 Add tests for `send_data_to_metax`

Co-authored-by: Kaisa Kuivalahti <kaisakui@users.noreply.github.com>
Co-authored-by: Sam Hardwick <Traubert@users.noreply.github.com>

kaisakui committed:
KP-7440 Tests sending data to metax when dict empty

kaisakui committed:
KP-7440 Add main() that handles pipeine

kaisakui committed:
KP-7440 Pass datetime instead of log file

kaisakui committed:
KP-7440 Add tests for main()

Traubert commented on tests/test_metadata_harvester_cli.py:
```diff
@@ -28,75 +10,239 @@ def kielipankki_api_url():
+
 
-    result = runner.invoke(retrieve_metadata_content, input=single_record_xml)
+def test_main_all_data_harvested(
```
We could make sure here that this test will run without an existing log file (when there's one locally).

kaisakui replied:
Refactored main() to have a log file as an argument. Changed the tests accordingly and added checking that the logger logs partial and complete harvests as expected with or without an existing date.

Traubert commented on tests/test_metadata_harvester_cli.py:
```diff
@@ -28,75 +10,239 @@ def kielipankki_api_url():
+    assert mock_requests_post.request_history[2].json()["persistent_identifier"] == list(single_record_to_dict.values())[0]["persistent_identifier"]
+
+
+def test_main_new_records_harvested_since_date(
```
And here we could make sure that something that isn't meant to be updated doesn't, and something that is is.

kaisakui replied:
Same comment here as above.

kaisakui committed:
KP-7440 Fix to_dict() tests to read from test log

kaisakui committed:
KP-7440 Remove urls and change call_counts

kaisakui committed:
KP-7440 Include checking partial/complete harvests

kaisakui committed:
KP-7440 Pass log file as argument in main()

aajarven commented on tests/test_metadata_harvester_cli.py:
```diff
@@ -28,75 +11,268 @@ def kielipankki_api_url():
+        "2023-09-08 14:45:58,690 - INFO - Started\n"
+        "2023-09-08 14:45:58,956 - INFO - Success, records harvested since 2023-09-08T14:34:16Z\n"
+    ]
+    with open(log_file, "w") as file:
+        file.writelines(log_file_data)
+    yield log_file
+    os.remove(log_file)
```
We don't need to take care of manual file creation and deletion in the current directory (which can cause problems like overwriting content or the test file not getting removed if the tests crash) if we use the pre-made [tmp_path](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html) fixture. With that we could do something like
```
log_file = tmp_path / "harvester_test.log"
log_file.writelines(log_file_data)
return log_file
```

The same goes for other places where we manually create files.

kaisakui committed:
KP-7440 Use tmp_path for test log files

kaisakui committed:
KP-7440 Remove redundant comment

aajarven reviewed (approved):
:+1:
